src/java/org/apache/cassandra/index/sai/utils/RangeIterator.java

```diff
@@ -0,0 +1,365 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.io.Closeable;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.PriorityQueue;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.utils.AbstractGuavaIterator;
+
+/**
+ * An abstract implementation of {@link AbstractGuavaIterator} that supports the building and management of
+ * concatanation, union and intersection iterators.
+ */
+public abstract class RangeIterator extends AbstractGuavaIterator<PrimaryKey> implements Closeable
+{
+    private final PrimaryKey min, max;
+    private final long count;
+    private PrimaryKey current;
+
+    protected RangeIterator(Builder.Statistics statistics)
+    {
+        this(statistics.min, statistics.max, statistics.tokenCount);
+    }
+
+    public RangeIterator(PrimaryKey min, PrimaryKey max, long count)
+    {
+        boolean isComplete = min != null && max != null && count != 0;
+        boolean isEmpty = min == null && max == null && (count == 0 || count == -1);
+        Preconditions.checkArgument(isComplete || isEmpty, "Range: [" + min + ',' + max + "], Count: " + count);
+
+        this.min = min;
+        this.current = min;
+        this.max = max;
+        this.count = count;
+    }
+
+    public final PrimaryKey getMinimum()
+    {
+        return min;
+    }
+
+    public final PrimaryKey getCurrent()
+    {
+        return current;
+    }
+
+    public final PrimaryKey getMaximum()
+    {
+        return max;
+    }
+
+    public final long getCount()
+    {
+        return count;
+    }
+
+    /**
+     * When called, this iterators current position should
+     * be skipped forwards until finding either:
+     *   1) an element equal to or bigger than next
+     *   2) the end of the iterator
+     *
+     * @param nextToken value to skip the iterator forward until matching
+     *
+     * @return The next current token after the skip was performed
+     */
+    public final PrimaryKey skipTo(PrimaryKey nextToken)
```

Comment:
nit/naming: I'm guessing `nextToken` is a hold-over from SASI's version of this. Perhaps `target`, `targetKey`, or just `key`. Same thing in the JavaDoc (ex. "The next current token")